File: BE_StayHub/app/Models/Tenant.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Notifications\Notifiable;
use Laravel\Sanctum\HasApiTokens;
use Laravel\Scout\Searchable;

class Tenant extends Authenticatable
{
    use HasApiTokens, HasFactory, Notifiable, Searchable;

    public const SEARCH_INDEX = 'tenants';

    public const GENDER_MALE = 1;
    public const GENDER_FEMALE = 2;

    public const GENDER_LABELS = [
        self::GENDER_MALE => 'Nam',
        self::GENDER_FEMALE => 'Nữ',
    ];

    public const STATUS_RENTING = 1;
    public const STATUS_STOPPED_RENTING = 2;

    public const STATUS_LABELS = [
        self::STATUS_RENTING => 'Đang thuê',
        self::STATUS_STOPPED_RENTING => 'Ngừng thuê',
    ];

    public const IDENTITY_TYPE_CCCD = 1;
    public const IDENTITY_TYPE_CMND = 2;
    public const IDENTITY_TYPE_PASSPORT = 3;

    public const IDENTITY_TYPE_LABELS = [
        self::IDENTITY_TYPE_CCCD => 'CCCD',
        self::IDENTITY_TYPE_CMND => 'CMND',
        self::IDENTITY_TYPE_PASSPORT => 'Hộ chiếu',
    ];

    protected $fillable = ['created_by', 'room_id', 'full_name', 'gender', 'date_of_birth', 'phone', 'email', 'username', 'password', 'permanent_address', 'current_address', 'avatar_url', 'status', 'identity_type', 'identity_number', 'front_image_url', 'back_image_url'];

    protected $hidden = ['password', 'remember_token'];

    protected function casts(): array
    {
        return [
            'created_by' => 'integer',
            'gender' => 'integer',
            'date_of_birth' => 'date',
            'status' => 'integer',
            'identity_type' => 'integer',
            'password' => 'hashed',
        ];
    }

    public function searchableAs(): string
    {
        return self::SEARCH_INDEX;
    }

    /**
     * Dữ liệu đồng bộ sang Meilisearch, không đưa mật khẩu/token/ảnh giấy tờ nhạy cảm vào index.
     */
    public function toSearchableArray(): array
    {
        return [
            'id' => $this->id,
            'created_by' => $this->created_by,
            'full_name' => $this->full_name,
            'username' => $this->username,
            'phone' => $this->phone,
            'email' => $this->email,
            'gender' => $this->gender === null ? null : (int) $this->gender,
            'status' => $this->status === null ? null : (int) $this->status,
            'identity_type' => $this->identity_type === null ? null : (int) $this->identity_type,
            'identity_number' => $this->identity_number,
            'permanent_address' => $this->permanent_address,
            'current_address' => $this->current_address,
            'date_of_birth' => optional($this->date_of_birth)->toDateString(),
            'created_at' => optional($this->created_at)->timestamp,
            'updated_at' => optional($this->updated_at)->timestamp,
        ];
    }

    public function creator(): BelongsTo
    {
        return $this->belongsTo(Admin::class, 'created_by');
    }

    public function room(): BelongsTo
    {
        return $this->belongsTo(Room::class);
    }

    public function representedContracts(): HasMany
    {
        return $this->hasMany(Contract::class, 'representative_tenant_id');
    }

    public function contractTenants(): HasMany
    {
        return $this->hasMany(ContractTenant::class);
    }

    public function contracts(): BelongsToMany
    {
        return $this->belongsToMany(Contract::class, 'contract_tenants')->withPivot(['join_date', 'leave_date', 'billing_start_date', 'billing_end_date', 'is_representative', 'is_staying', 'created_by'])->withTimestamps();
    }

    public function roomMovements(): HasMany
    {
        return $this->hasMany(RoomMovement::class);
    }

    public function vehicles(): HasMany
    {
        return $this->hasMany(Vehicle::class);
    }

    public function maintenanceRequests(): HasMany
    {
        return $this->hasMany(MaintenanceRequest::class);
    }

    public function maintenanceFeedbacks(): HasMany
    {
        return $this->hasMany(MaintenanceFeedback::class);
    }

    public function notificationReads(): HasMany
    {
        return $this->hasMany(NotificationRead::class);
    }

    public function readNotifications(): BelongsToMany
    {
        return $this->belongsToMany(Notification::class, 'notification_reads')->withPivot('read_at');
    }
}

File: FE_StayHub_Mobile/lib/models/tenant.dart
class Tenant {
  final int id;
  final int? buildingId;
  final String fullName;
  final int gender;
  final String? dateOfBirth;
  final String phone;
  final String email;
  final String username;
  final String? permanentAddress;
  final String? currentAddress;
  final String? avatarUrl;
  final int status; // 1: Renting, 2: Stopped Renting
  final String? roomNumber;
  final String? buildingName;
  final int identityType;
  final String identityNumber;
  final String? frontImageUrl;
  final String? backImageUrl;

  Tenant({
    required this.id,
    this.buildingId,
    required this.fullName,
    required this.gender,
    this.dateOfBirth,
    required this.phone,
    required this.email,
    required this.username,
    this.permanentAddress,
    this.currentAddress,
    this.avatarUrl,
    required this.status,
    this.roomNumber,
    this.buildingName,
    required this.identityType,
    required this.identityNumber,
    this.frontImageUrl,
    this.backImageUrl,
  });

  factory Tenant.fromJson(Map<String, dynamic> json) {
    return Tenant(
      id: json['id'] as int,
      buildingId: json['building_id'] as int?,
      fullName: json['full_name'] as String,
      gender: json['gender'] as int? ?? 1,
      dateOfBirth: json['date_of_birth'] as String?,
      phone: json['phone'] as String? ?? '',
      email: json['email'] as String? ?? '',
      username: json['username'] as String? ?? '',
      permanentAddress: json['permanent_address'] as String?,
      currentAddress: json['current_address'] as String?,
      avatarUrl: json['avatar_url'] as String?,
      status: json['status'] as int? ?? 1,
      roomNumber: json['room_number'] as String?,
      buildingName: json['building_name'] as String?,
      identityType: json['identity_type'] as int? ?? 1,
      identityNumber: json['identity_number'] as String? ?? '',
      frontImageUrl: json['front_image_url'] as String?,
      backImageUrl: json['back_image_url'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'building_id': buildingId,
      'full_name': fullName,
      'gender': gender,
      'date_of_birth': dateOfBirth,
      'phone': phone,
      'email': email,
      'username': username,
      'permanent_address': permanentAddress,
      'current_address': currentAddress,
      'avatar_url': avatarUrl,
      'status': status,
      'room_number': roomNumber,
      'building_name': buildingName,
      'identity_type': identityType,
      'identity_number': identityNumber,
      'front_image_url': frontImageUrl,
      'back_image_url': backImageUrl,
    };
  }

  String get genderLabel => gender == 1 ? 'Nam' : 'Nữ';
  String get statusLabel => status == 1 ? 'Đang thuê' : 'Ngừng thuê';
  bool get isActive => status == 1;
}
